### PR TITLE
Add 2D case for at::convolution

### PR DESCRIPTION
**Pull request by gpetters94**
Here's the implementation of `at::convolution` and a decomposition of `at::convolution_overrideable`. The 2D case is all that's implemented for `at::convolution`, since the 1D and 3D cases require changes to linalg. I don't think we need them right now, looking at our op list, but I need to look further into the list to be sure. `convolution_overrideable` just decomposes directly into `convolution`.

### CODE REVIEW

**Comment by cathyzhyi** [comment]
The code seems to be very similar to conv2d. Can we just decompose `aten::convolution ` to `aten::conv2d` when it's 2D?

**Comment by gpetters94** [comment]
We could do that, yeah. Though if we do end up adding 1D and 3D cases it might make more sense to decompose conv1d/conv2d/conv3d to `aten::convolution`.

**Comment by cathyzhyi** [comment]
> We could do that, yeah. Though if we do end up adding 1D and 3D cases it might make more sense to decompose conv1d/conv2d/conv3d to `aten::convolution`.

yea, that works too. Can you decompose conv2d to `aten::convolution` to reduce duplication?

File: e2e_testing/torchscript/conv.py [hunk 1]
```diff
@@ -133,3 +133,59 @@ def forward(self, x):
...
+        super().__init__()
+
+    @export
+    @annotate_args([
+        None,
+        ([-1, -1, -1, -1], torch.float32, True),
+        ([-1, -1, -1, -1], torch.float32, True),
+    ])
+    def forward(self, inputVec, weight):
+        return torch.ops.aten.convolution(inputVec, weight, bias=None, stride=[1, 1], padding=[0, 0], dilation=[1, 1], transposed=False, output_padding=[0, 0], groups=1)
```
Comment (cathyzhyi) [inline e2e_testing/torchscript/conv.py:168]: This line is too long. can you change the argument format to be the same as the other tests?

File: e2e_testing/torchscript/conv.py [hunk 2]
```diff
@@ -133,3 +133,59 @@ def forward(self, x):
...
+
+@register_test_case(module_factory=lambda: ConvolutionModule2D())
+def ConvolutionModule2D_basic(module, tu: TestUtils):
+    with torch.backends.mkldnn.flags(enabled=False):
```
Comment (cathyzhyi) [inline e2e_testing/torchscript/conv.py:172]: why is this line needed?

File: e2e_testing/torchscript/conv.py [hunk 3]
```diff
@@ -133,3 +133,59 @@ def forward(self, x):
...
+        module.forward(torch.ones(3, 3, 10, 10), torch.ones(3, 3, 2, 2))
```
Comment (cathyzhyi) [inline e2e_testing/torchscript/conv.py:173]: can we use more random data rather than all ones?

File: lib/Conversion/TorchToLinalg/Linear.cpp [hunk 4]
```diff
@@ -465,36 +465,40 @@ class ConvertAtenConv2dOp : public OpConversionPattern<AtenConv2dOp> {
         input.getType().cast<RankedTensorType>().getElementType();
     if (!elementType.isa<mlir::FloatType>())
       return op.emitError("unimplemented: non-floating point type");
+    SmallVector<Value> inSizes = getTensorSizes(rewriter, loc, input);
```
Comment (cathyzhyi) [inline lib/Conversion/TorchToLinalg/Linear.cpp:468]: We can get the rank directly from input ranked tensor type.

File: lib/Conversion/TorchToLinalg/Linear.cpp [hunk 5]
```diff
@@ -465,36 +465,40 @@ class ConvertAtenConv2dOp : public OpConversionPattern<AtenConv2dOp> {
...
+    size_t inRank = inSizes.size();
+    if (inRank != 4)
+      return rewriter.notifyMatchFailure(
+          op, "only 2D convolution currently supported");
```
Comment (cathyzhyi) [inline lib/Conversion/TorchToLinalg/Linear.cpp:472]: nit: add `unimplemented:` to be consistent with other error msg

File: lib/Conversion/TorchToLinalg/Linear.cpp [hunk 6]
```diff
@@ -559,14 +562,32 @@ class ConvertAtenConv2dOp : public OpConversionPattern<AtenConv2dOp> {
 
     auto stridesAttr = rewriter.getI64VectorAttr(strideInts);
     auto dilationAttr = rewriter.getI64VectorAttr(dilationInts);
-    Value conv2d =
+    Value conv;
```
Comment (cathyzhyi) [inline lib/Conversion/TorchToLinalg/Linear.cpp:565]: can you remove the commented out code and add a // TODO:  add support for 2d and 2d case

File: python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/torch_ods_gen.py [hunk 7]
```diff
@@ -27,6 +27,7 @@
     "int": "Torch_IntType",
     "int[]": "TorchIntListType",
     "int?": "TorchOptionalIntType",
+    "int[]?": "TorchOptionalIntListType",
```
Comment (cathyzhyi) [inline python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/torch_ods_gen.py:30]: Can you fix this in the other PR I mentioned before? https://github.com/llvm/torch-mlir/pull/643

**Review by cathyzhyi — CHANGES_REQUESTED** [review]

**Comment by cathyzhyi** [comment]
Can you help to fix TOSA as well here? https://github.com/llvm/torch-mlir/blob/5b9bdfaf3fb56c5aed12d689e21c9884b1c5a789/lib/Conversion/TorchToTosa/TorchToTosa.cpp#L1681. Just need to change the conversion pattern to apply for `aten::convolution` and guard `inputRank == 4` as we do in Torch to Linalg conversion.

File: lib/Dialect/Torch/Transforms/ShapeLibrary.cpp [hunk 8]
```diff
@@ -39,11 +39,11 @@ module {
   func @__torch__.torch_mlir.dialects.torch.importer.jit_ir.build_tools.upstream_shape_helpers._copy(%arg0: !torch.list<int>) -> !torch.list<int> {
     %true = torch.constant.bool true
     %0 = torch.prim.ListConstruct  : () -> !torch.list<int>
-    %1 = torch.aten.len.t %arg0 : !torch.list<int> -> !torch.int
+    %1 = torch.aten.len.t %arg0 :  !torch.list<int> ->  !torch.int
```
Comment (cathyzhyi) [inline lib/Dialect/Torch/Transforms/ShapeLibrary.cpp:42]: Can you rebase to the latest and rerun `update_shape_lib.sh`. A recent fix from https://github.com/llvm/torch-mlir/pull/687 should avoid adding the extra spaces.

File: e2e_testing/torchscript/conv.py (hunk 2 above)
Comment (gpetters94) [reply e2e_testing/torchscript/conv.py:188]: We've had some issues with running certain e2e tests because the mkldnn backend doesn't support some ops. I think convolution was one of those, but it may have been convolution_backward.

File: python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/torch_ods_gen.py (hunk 7 above)
Comment (gpetters94) [reply python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/torch_ods_gen.py:30]: Do you mind if I change it in this PR?

**Comment by gpetters94** [comment]
Re:TOSA, I've added the requested changes and it is still failing. TOSA is out of my wheelhouse, so that's the best I can do.

**Comment by cathyzhyi** [comment]
> Re:TOSA, I've added the requested changes and it is still failing. TOSA is out of my wheelhouse, so that's the best I can do.

The failure is at https://github.com/llvm/torch-mlir/blob/5b9bdfaf3fb56c5aed12d689e21c9884b1c5a789/lib/Conversion/TorchToTosa/TorchToTosa.cpp#L1715. This is because TOSA doesn't handle static weight. Can you add a test with static weight?

**Comment by sjarus** [comment]
I think you mean doesn't handle dynamic weight, @cathyzhyi  ? Yes the way around is to ensure the weight is static . Please see examples of static test versions that work for TOSA, e.f. vision_models.py has a ResNet18Static that works , while the original dynamic one doesn't yet.

**Comment by cathyzhyi** [comment]
> I think you mean doesn't handle dynamic weight, @cathyzhyi ? Yes the way around is to ensure the weight is static . Please see examples of static test versions that work for TOSA, e.f. vision_models.py has a ResNet18Static that works , while the original dynamic one doesn't yet.

yea.. sry, I meant dynamic weight. Thanks for confirming!

**Comment by gpetters94** [comment]
I think the TOSA fix is working, but I'm not sure how to mark the test as an expected pass for the TOSA pass.

**Comment by cathyzhyi** [comment]
you can the test to this list https://github.com/llvm/torch-mlir/blob/1b505cbac506d80a466f88afc4d0c6a6aa3d347c/e2e_testing/torchscript/xfail_sets.py#L25

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 9]
```diff
@@ -1701,6 +1701,12 @@ LogicalResult ConvertAtenOp<AtenConvolutionOp>::matchAndRewrite(
 
   if (inputTy.getRank() != 4)
     return op.emitError("Unimplemented: only 2D convolutions supported");
+  
+  bool staticWeight = true;
```
Comment (cathyzhyi) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:1705]: nit: you can use `hasStaticShape` on RankedTensorType to check for static shape

File: python/torch_mlir_e2e_test/test_suite/conv.py [hunk 10]
```diff
@@ -213,3 +213,29 @@ def ConvolutionModule2D_basic(module, tu: TestUtils):
...
+                                          padding=[0, 0],
+                                          dilation=[1, 1],
+                                          transposed=False,
+                                          output_padding=[0, 0],
+                                          groups=1)
+
+@register_test_case(module_factory=lambda: ConvolutionModule2DStatic())
+def ConvolutionModule2DStatic_basic(module, tu: TestUtils):
+    with torch.backends.mkldnn.flags(enabled=False):
+        module.forward(torch.randn(3, 3, 10, 10), torch.randn(3, 3, 2, 2))
```
Comment (cathyzhyi) [inline python/torch_mlir_e2e_test/test_suite/conv.py:241]: nit: please add an extra empty line at the end of the file.

**Review by cathyzhyi — CHANGES_REQUESTED** [review]

File: lib/Conversion/TorchToTosa/TorchToTosa.cpp [hunk 11]
```diff
@@ -1692,13 +1692,19 @@ LogicalResult ConvertAtenOp<AtenConv2dOp>::matchAndRewrite(
...
   auto inputElemTy = inputTy.getElementType();
   auto weightElemTy = weightTy.getElementType();
   auto inputShape = inputTy.getShape();
   auto weightShape = weightTy.getShape();
 
+  if (inputTy.getRank() != 4)
+    return op.emitError("Unimplemented: only 2D convolutions supported");
+
+  if (!weightTy.hasStaticShape())
+    return op->emitError("Unimplemented: TOSA only supports static weight");
```
Comment (cathyzhyi) [inline lib/Conversion/TorchToTosa/TorchToTosa.cpp:1706]: nit: use `op.` for consistency.

File: python/torch_mlir_e2e_test/test_suite/conv.py [hunk 12]
```diff
@@ -133,3 +133,109 @@ def forward(self, x):
 def Conv2dWithPaddingDilationStrideStaticModule_basic(module, tu: TestUtils):
     t = tu.rand(5, 2, 10, 20)
     module.forward(t)
+
+# ==============================================================================
+
+# class ConvolutionModule1D(torch.nn.Module):
```
Comment (cathyzhyi) [inline python/torch_mlir_e2e_test/test_suite/conv.py:139]: those tests could be uncommented and added into xfail_set.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 13]
```diff
@@ -750,6 +750,47 @@ class DecomposeAtenWhereScalarSelfOp
...
+class DecomposeAtenConv2dOp : public OpRewritePattern<AtenConv2dOp> {
+public:
+  using OpRewritePattern::OpRewritePattern;
+  LogicalResult matchAndRewrite(AtenConv2dOp op,
+                                PatternRewriter &rewriter) const override {
+
+    Value emptyList = rewriter.create<PrimListConstructOp>(
+        op.getLoc(), Torch::ListType::get(Torch::IntType::get(op.getContext())),
+        SmallVector<Value>());
+    Value False = rewriter.create<Torch::ConstantBoolOp>(op.getLoc(), false);
```
Comment (cathyzhyi) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:783]: nit: use `cstFalse` to follow the naming convention.

File: python/torch_mlir_e2e_test/test_suite/conv.py [hunk 14]
```diff
@@ -133,3 +133,109 @@ def forward(self, x):
...
+    @annotate_args([
+        None,
+        ([-1, -1, -1, -1], torch.float32, True),
+        ([-1, -1, -1, -1], torch.float32, True),
+    ])
+    def forward(self, inputVec, weight):
+        return torch.ops.aten.convolution(inputVec,
+                                          weight,
+                                          bias=None,
+                                          stride=[1, 1],
```
Comment (cathyzhyi) [inline python/torch_mlir_e2e_test/test_suite/conv.py:179]: shall we change the strides to values other than 1 and padding to non zero?

**Review by cathyzhyi — CHANGES_REQUESTED** [review]
`'aten::convolution` has more arguments than `aten::conv2d` like `transposed` and `output_padding`. If we don't support those arguments we need to guard against the values that are not supported.

**Comment by gpetters94** [comment]
I've added a guard for transposed. As I understand it output_padding is just a compiler hint, so we shouldn't need to worry about it.

File: lib/Conversion/TorchToLinalg/Linear.cpp [hunk 15]
```diff
@@ -465,42 +465,53 @@ class ConvertAtenConv2dOp : public OpConversionPattern<AtenConv2dOp> {
...
+    Value N = getDimOp(rewriter, loc, input, 0);
+    SmallVector<Value> inDims;
+    for (size_t i = 2; i < inRank; i++)
+      inDims.push_back(getDimOp(rewriter, loc, input, i));
+    Value F = getDimOp(rewriter, loc, weight, 0);
+    SmallVector<Value> weightDims;
+    for (size_t i = 2; i < inRank; i++)
+      weightDims.push_back(getDimOp(rewriter, loc, weight, i));
+
+    // Guard unused values (transposed, groups)
```
Comment (cathyzhyi) [inline lib/Conversion/TorchToLinalg/Linear.cpp:501]: Can you change both guards to be consistent with other guards in the codebase using `!matchPattern`?

**Review by cathyzhyi — CHANGES_REQUESTED** [review]

**Comment by gpetters94** [comment]
Okay, the guards are both `matchPattern`s now.

File: lib/Conversion/TorchToLinalg/Linear.cpp [hunk 16]
```diff
@@ -465,42 +465,48 @@ class ConvertAtenConv2dOp : public OpConversionPattern<AtenConv2dOp> {
...
+    int64_t group_size;
+    if(!matchPattern(op.groups(), m_TorchConstantInt(&group_size)) || group_size != 1)
+      return rewriter.notifyMatchFailure(op,
+                                         "unimplemented: only group size of 1 supported");
+    bool transposed;
+    if(!matchPattern(op.transposed(), m_TorchConstantBool(&transposed)))
```
Comment (cathyzhyi) [inline lib/Conversion/TorchToLinalg/Linear.cpp:507]: also need to check `transposed` is false

File: python/torch_mlir_e2e_test/test_suite/conv.py (hunk 10 above)
Comment (cathyzhyi) [reply python/torch_mlir_e2e_test/test_suite/conv.py:267]: this is not addressed yet.

**Review by cathyzhyi — CHANGES_REQUESTED** [review]

File: lib/Conversion/TorchToLinalg/Linear.cpp (hunk 16 above)
Comment (gpetters94) [reply lib/Conversion/TorchToLinalg/Linear.cpp]: Good point, I've added that check.

Comment (cathyzhyi) [reply lib/Conversion/TorchToLinalg/Linear.cpp]: `bool transposed`  need to be initialized with `true`. Otherwise, when `op.transposed` is variable and `transposed` is initialized with `false`, the guard won't catch the case correctly.

File: e2e_testing/torchscript/conv.py (hunk 2 above)
Comment (cathyzhyi) [reply e2e_testing/torchscript/conv.py]: I missed this the last time. This code shouldn't be part of our e2e test right? because we don't use mkldnn backend in our e2e testing.

**Review by cathyzhyi — CHANGES_REQUESTED** [review]

File: python/torch_mlir_e2e_test/test_suite/conv.py [hunk 17]
```diff
@@ -133,3 +133,135 @@ def forward(self, x):
...
+                                          stride=[3, 3],
+                                          padding=[2, 2],
+                                          dilation=[1, 1],
+                                          transposed=False,
+                                          output_padding=[0, 0],
+                                          groups=1)
+
+@register_test_case(module_factory=lambda: ConvolutionModule2DStrided())
+def ConvolutionModule2DStrided_basic(module, tu: TestUtils):
+    with torch.backends.mkldnn.flags(enabled=False):
```
[the remark this answers is not in the document]
Comment (gpetters94) [reply python/torch_mlir_e2e_test/test_suite/conv.py]: Yep, it was convolution_backward. I've updated accordingly.

**Review by cathyzhyi — APPROVED** [review]
